File: ios/RailwayWidget/Domain/UseCase/GetPickerStationsUseCase.swift
struct GetPickerStationsUseCase {
    private let repository: IPickerStationRepository

    init(repository: IPickerStationRepository) {
        self.repository = repository
    }

    /// Returns up to 10 ordered picker stations for `system`.
    /// Defaults are seeded by the Flutter app on first launch (Dart/Drift).
    /// Falls back to in-memory defaults if the DB is unavailable or empty.
    func execute(system: String) -> [PickerStation] {
        do {
            let dbStations = try repository.getStations(system: system)
            guard !dbStations.isEmpty else {
                return PickerStationDefaults.stations(for: system)
            }
            if dbStations.count < 10 {
                let existingNames = Set(dbStations.map { $0.name })
                let extras = PickerStationDefaults.stations(for: system)
                    .filter { !existingNames.contains($0.name) }
                return Array((dbStations + extras).prefix(10))
            }
            return dbStations
        } catch {
            return PickerStationDefaults.stations(for: system)
        }
    }
}

File: android/app/src/main/kotlin/com/example/flutter_railway_timetable/widget/domain/usecase/GetPickerStationsUseCase.kt
package com.example.flutter_railway_timetable.widget.domain.usecase

import com.example.flutter_railway_timetable.widget.domain.entity.PickerStation
import com.example.flutter_railway_timetable.widget.domain.entity.PickerStationDefaults
import com.example.flutter_railway_timetable.widget.domain.repository.IWidgetStationRepository

// Mirrors iOS GetPickerStationsUseCase: returns up to 10 ordered picker stations,
// falling back to / topping up with PickerStationDefaults when the DB is missing
// rows or unavailable.
class GetPickerStationsUseCase(private val repository: IWidgetStationRepository) {

    suspend fun execute(system: String): List<PickerStation> {
        return try {
            val db = repository.getStations(system)
            when {
                db.isEmpty() -> PickerStationDefaults.stations(system)
                db.size < MAX_STATIONS -> {
                    val existingNames = db.map { it.name }.toSet()
                    val extras = PickerStationDefaults.stations(system)
                        .filter { it.name !in existingNames }
                    (db + extras).take(MAX_STATIONS)
                }
                else -> db
            }
        } catch (_: Exception) {
            PickerStationDefaults.stations(system)
        }
    }

    private companion object {
        const val MAX_STATIONS = 10
    }
}
